### ISSUE TITLE
[React 19] aria attributes on custom elements behave maybe incorrectly

### ISSUE THREAD

**Issue by epmatsw** [facebook/react#32251]
## Summary

In [React 18](https://codesandbox.io/p/sandbox/gracious-flower-dwt3cd), passing `aria-disabled={true}` to a custom element serialized that value as "true". That seemed good and as far as I can tell worked as expected.

In [React 19](https://codesandbox.io/p/sandbox/intelligent-dawn-mv9p62), that same code adds the `aria-disabled` attribute, but doesn't set it to anything. That doesn't work as well. It appears to me (although I'm not certain) that it breaks accessibility for these custom elements. It definitely breaks `@testing-library/jest-dom`'s `toBeChecked` matcher with `aria-checked` (https://github.com/testing-library/jest-dom/blob/918b6fbcde10d4409ee8f05c6e4eecbe96a72b7a/src/to-be-checked.js#L31).

For these aria attributes that require strings instead of booleans to function properly, is this an intended change?

**Comment by epmatsw** [comment]
Still an issue in 19.1

**Comment by mrginglymus** [comment]
I've just hit this with `aria-hidden` - on standard DOM nodes this comes through as `aria-hidden="true"`; on custom elements it comes out as `aria-hidden` (which is the same as not specifying it at all).

```html
<div aria-hidden  />
<my-element aria-hidden></my-element>
```

is rendered as

```html
<div aria-hidden="true"></div>
<my-element aria-hidden></my-element>
```

**Comment by epmatsw** [comment]
I'm pretty sure the issue is here:

https://github.com/facebook/react/blob/3958d5d84b3d3e6ae5c1caef9c8cf0dc58e862e6/packages/react-dom-bindings/src/client/DOMPropertyOperations.js#L228-L231

That `if (value === true)` can maybe be `if (value === true && !name.toLowerCase().startsWith("aria-"))`.

**Comment by mrginglymus** [comment]
We're also missing handling for e.g. `aria-selected="false"` - which is [different from omitting it](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-selected#false).

```
<div aria-hidden={false} />
<my-element aria-hidden={false}></my-element>
```

is rendered as

```
<div aria-hidden="false"></div>
<my-element></my-element>
```

**Comment by mrginglymus** [comment]
> I'm pretty sure the issue is here:
> 
> [react/packages/react-dom-bindings/src/client/DOMPropertyOperations.js](https://github.com/facebook/react/blob/3958d5d84b3d3e6ae5c1caef9c8cf0dc58e862e6/packages/react-dom-bindings/src/client/DOMPropertyOperations.js#L228-L231)
> 
> Lines 228 to 231 in [3958d5d](/facebook/react/commit/3958d5d84b3d3e6ae5c1caef9c8cf0dc58e862e6)
> 
>  if (value === true) { 
>    node.setAttribute(name, ''); 
>    return; 
>  } 
> That `if (value === true)` can maybe be `if (value === true && !name.toLowerCase().startsWith("aria-"))`.

Should probably check for `data-` attrs too for parity with native elements.

Seems to be a similar bug in `getValueForAttributeOnCustomComponent`, where it'll return `false` instead of `undefined`.

**Comment by mrginglymus** [comment]
bump

**Comment by epmatsw** [comment]
Bump

**Comment by MaxwellCohen** [comment]
This is still an issue. I have created a basic recreation here: https://codesandbox.io/p/sandbox/react-dev-forked-www8qx using the latest stable version of React. 

aria-hidden attribute values "undefined" and "true" mean diffrent things and must be strings. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-hidden#values): 

> [false](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-hidden#false)
The element is exposed to the accessibility API as if it was rendered.
> [true](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-hidden#true)
The element is hidden from the accessibility API.
> [undefined (default)](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-hidden#undefined)
The element's hidden state is determined by the user agent based on whether it is rendered.

These differences are passed to the elements API so web components can use them. The workaround is to use string values instead of booleans, but it takes away some of the niceness of JSX.

**Comment by MaxwellCohen** [comment]
I have been looking into creating a solution for this issue. The current React 19 behavior is a bug (in my opinion), but the change could be considered a breaking change. I also discovered we are having the same issue with data-* attributes. 

The workaround is to do the following
```
<my-element aria-hidden={isHidden ? 'true' : 'false'}></my-element>
```

**Comment by MaxwellCohen** [comment]
I have created a PR to try to fix this issue. Any feedback would be greatly appreciated. 

You can test it by using the preview branch version of React.
```
# yarn 1
yarn add https://pkg.csb.dev/facebook/react/commit/97b60066/react https://pkg.csb.dev/facebook/react/commit/97b60066/react-dom https://pkg.csb.dev/facebook/react/commit/97b60066/react-server-dom-webpack https://pkg.csb.dev/facebook/react/commit/97b60066/scheduler
# yarn 2, 3
yarn add react@https://pkg.csb.dev/facebook/react/commit/97b60066/react/_pkg.tgz react-dom@https://pkg.csb.dev/facebook/react/commit/97b60066/react-dom/_pkg.tgz react-server-dom-webpack@https://pkg.csb.dev/facebook/react/commit/97b60066/react-server-dom-webpack/_pkg.tgz scheduler@https://pkg.csb.dev/facebook/react/commit/97b60066/scheduler/_pkg.tgz
# npm
npm i https://pkg.csb.dev/facebook/react/commit/97b60066/react https://pkg.csb.dev/facebook/react/commit/97b60066/react-dom https://pkg.csb.dev/facebook/react/commit/97b60066/react-server-dom-webpack https://pkg.csb.dev/facebook/react/commit/97b60066/scheduler
```

**Comment by otomad** [comment]
Bump.
